### PR TITLE
Merge advanced search interface and search results into a single page

### PR DESCRIPTION
**Pull request by srisi**
i.e. a user executing a query now will get the results displayed below the query.
Also:
- Added metadata information to each search result (this should be further expanded though.)
- Removed all of the try...except blocks from the advanced search.
- Removed filtering by page number. I think few queries need filters by document length.
- Tested with various different text/author/recipient queries. Full text search does not return proper results for fixed expressions, e.g. "course 6." That returns documents mentioning course and 6 separately. Still, I think this provides a good basis for someone else to continue work on the search.

@angcast: There is open space on the right side of the search results to display the facet results.

### CODE REVIEW

File: djweb/dj_comp_hist/views.py [hunk 1]
```diff
@@ -196,101 +198,83 @@ def browse(request):
...
+            print(f"WARNING. Could not parse full text search string: {search_params['text']}.")
+        else:
+            raw_docs = Document.objects.raw(f'''SELECT * FROM doc_fts 
+                                                     WHERE text MATCH "{match.groups()[0]}";''')
             doc_ids = [doc.id for doc in raw_docs]
             doc_objs = doc_objs.filter(id__in=doc_ids)
 
-    except:
-        print("Error getting phrase")
+    if search_params['author'] != '':
```
Comment (ryaanahmed) [inline djweb/dj_comp_hist/views.py]: @srisi does this author filtering logic seem a bit off to you? Why are we only checking the first word in the query against all of the author fields?

File: djweb/dj_comp_hist/views.py [hunk 2]
```diff
@@ -196,101 +198,83 @@ def browse(request):
...
-    except:
-        print('Error getting doc types')
-    try:
-        pages = [int(request.GET['minPages']), int(request.GET['maxPages'])]
-        doc_objs = doc_objs.filter(Q(number_of_pages__gte=pages[0]) &
-                                   Q(number_of_pages__lte=pages[1]))
-    except:
-        print('Error getting pages')
+    if search_params['recipient'] != '':
+        recipient = search_params['recipient'].split(" ")
```
Comment (ryaanahmed) [inline djweb/dj_comp_hist/views.py]: Same question here as for author.  Seems like what we want are queries like `Q(recipient_person__first__in=recipient`. Trouble is this is not case insensitive. and there's no __iin=. Hmm.

File: djweb/dj_comp_hist/views.py (hunk 1 above)
Comment (srisi) [reply djweb/dj_comp_hist/views.py]: Agreed. I had copied an earlier version yesterday. However, on second thought, there is no reason why we shouldn't also use the case insensitive full text search for title, author, and recipient. -> updated.

File: djweb/dj_comp_hist/metadata_parser.py [hunk 3]
```diff
@@ -56,15 +56,34 @@ def populate_from_metadata(file_name=None):
...
+    cursor.execute('''CREATE VIRTUAL TABLE doc_fts USING FTS4(id, title, text, author, recipient, 
+                                                                                       cced);''')
+    for d in Document.objects.all():
+        # join together all author names. By default, first and last name are joined by an
+        # underscore -> replace with a space
+        author = " ".join([p['name'].replace('_', ' ') for p in d.get_person_list('authors')])
+        recipient = " ".join([p['name'].replace('_', ' ') for p in d.get_person_list('recipients')])
+        cced = " ".join([p['name'].replace('_', ' ') for p in d.get_person_list('cceds')])
+
+        # Technically, sqlite expects quotation marks and apostrophes to be escaped by doubling
```
Comment (ryaanahmed) [inline djweb/dj_comp_hist/metadata_parser.py]: There were escaping problems in some of the other fields, so I'm applying this same removal of ' and " to all the fields as a temporary workaround.

**Comment by ryaanahmed** [comment]
Definitely a step in the right direction, as your version actually returns correct results!

However, I don't love that we're writing raw SQLite commands at this stage: it isn't robust to changes in the database engine, and I don't imagine we'll want to use SQLite in production. We can write this same query in the Django ORM by dynamically generating a Q-object, which will work regardless of DB backend. If we find this is too slow later on, we can optimize with a backend-specific command later, but let's not get locked into something like that now. I'd suggest that we rewrite full text search using this method, too, but deferring that for now.

A few other things I've changed:

- request.GET is a querydict, which is a dict subclass, so we don't need to call .dict().

- Checking for values against the empty string is not robust to queries where we leave out a search param completely (which we can imagine doing if, e.g., we make an API for this down the road, or if a clever user realizes they can manually edit the URL to pass different search queries).

- - e.g. `advanced_search/?title=foo` will crash when search_params['text'] doesn't find a matching key.

- - Using dict.get() makes this all a lot more robust, as it'll return None if the key isn't set, and you can just check the truthiness of the value it returns to decide whether to add the filter or not (since both None and '' are falsey).

- prefetch_related grabs the related objects that we use in the template, so we're not pinging the database per search result when calling render. speeds things up a lot.

**Comment by srisi** [comment]
Very good changes. I look forward to learning more about django internals and the proper ways to use them in the next week!
One final note: I don't think the filters are applied.
`docs_qs.filter(author_q)` should presumably be `docs_qs = docs_qs.filter(author_q)`
The same applies to recipients and doc_types. Currently, searching for author=huskey (1 matching doc) returns about 300 hits.

**Comment by ryaanahmed** [comment]
> One final note: I don't think the filters are applied.
Meep! You're completely right. I'll hotfix this immediately.

**Pull request merged**